src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,886 @@
+
+    def __call__(self, val):
+        return fq_default(val, self)
+
+
+cdef class fq_default(flint_scalar):
+    def __cinit__(self, val, ctx):
+        if not typecheck(ctx, fq_default_ctx):
+            raise TypeError
+        self.ctx = ctx
+        fq_default_init(self.val, self.ctx.val)
+
+    def __dealloc__(self):
+        if self.ctx is not None:
+            fq_default_clear(self.val, self.ctx.val)
+
+    def __init__(self, val, ctx):
+        if not typecheck(ctx, fq_default_ctx):
+            raise TypeError
+        self.ctx = ctx
+
+        # Converts the list to an fmpz_poly and then sets from this
+        if typecheck(val, list):
+            self.ctx.set_list_as_fq_default(self.val, val)
+            return
+
+        # Otherwise cascades through types to convert
+        check = self.ctx.set_any_as_fq_default(self.val, val)
+        if check is NotImplemented:
+            raise TypeError
+
+    def __int__(self):
+        """
+        Attempts to lift self to an integer of type fmpz in [0, p-1]
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> int(gf(123))
+            123
+        """
+        cdef fmpz x = fmpz.__new__(fmpz)
+        res = fq_default_get_fmpz(x.val, self.val, self.ctx.val)
+        if res == 1:
+            return int(x)
+        raise ValueError("fq element has no lift to the integers")
+
+    def polynomial(self):
+        """
+        Returns a representative of ``self`` as a polynomial in `(Z/pZ)[x] / h(x)`
+        where `h(x)` is the defining polynomial of the finite field.
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> a = gf([1,2,3])
+            >>> a.polynomial()
+            3*x^2 + 2*x + 1
+            >>> gf(123).polynomial()
+            123
+        """
+        cdef fmpz_mod_poly pol
+
+        ring_ctx = fmpz_mod_poly_ctx(self.ctx.prime())
+        pol = ring_ctx.new_ctype_poly()
+        fq_default_get_fmpz_mod_poly((<fmpz_mod_poly>pol).val, self.val, self.ctx.val)
+
+        return pol
+
+    def to_list(self):
+        """
+        Returns self as a list of fmpz types corresponding to a
+        list of coefficients
+
+            >>> gf = fq_default_ctx(163, 3)
+            >>> gf([-1,2,1]).to_list()
+            [162, 2, 1]
+            >>> gf.one().to_list()
+            [1, 0, 0]
+        """
+        coeffs = [None for _ in range(self.ctx.degree())]
+        for i in range(self.ctx.degree()):
+            c = fmpz.__new__(fmpz)
+            fq_default_get_coeff_fmpz((<fmpz>c).val, self.val, <slong>i, self.ctx.val)
+            coeffs[i] = c
+        return coeffs
+
+    def str(self):
+        return self.polynomial().str(var=self.ctx.var.decode())
+
+    def repr(self):
+        return f"fq_default({self.to_list(), self.ctx.__repr__()})"
+
+    def __hash__(self):
+        return hash((self.to_polynomial(), hash(self.ctx)))
```

Comment:
I think we probably need coverage measurement to make sure that cases like this are checked:
```python
In [69]: hash(F(4))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[69], line 1
----> 1 hash(F(4))

File fq_default.pyx:556, in flint.types.fq_default.fq_default.__hash__()

AttributeError: 'flint.types.fq_default.fq_default' object has no attribute 'to_polynomial'
```